flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalog.java

```diff
@@ -0,0 +1,259 @@
+	public GenericHiveMetastoreCatalog(String catalogName, String hivemetastoreURI) {
+		this(catalogName, getHiveConf(hivemetastoreURI));
+	}
+
+	public GenericHiveMetastoreCatalog(String catalogName, HiveConf hiveConf) {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(catalogName), "catalogName cannot be null or empty");
+		this.catalogName = catalogName;
+
+		this.hiveConf = checkNotNull(hiveConf, "hiveConf cannot be null");
+		LOG.info("Created GenericHiveMetastoreCatalog '{}'", catalogName);
+	}
+
+	private static HiveConf getHiveConf(String hiveMetastoreURI) {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(hiveMetastoreURI), "hiveMetastoreURI cannot be null or empty");
+
+		HiveConf hiveConf = new HiveConf();
+		hiveConf.setVar(HiveConf.ConfVars.METASTOREURIS, hiveMetastoreURI);
+		return hiveConf;
+	}
+
+	private static IMetaStoreClient getMetastoreClient(HiveConf hiveConf) {
+		try {
+			return RetryingMetaStoreClient.getProxy(
+				hiveConf,
+				null,
+				null,
+				HiveMetaStoreClient.class.getName(),
+				true);
+		} catch (MetaException e) {
+			throw new CatalogException("Failed to create Hive metastore client", e);
+		}
+	}
+
+	@Override
+	public void open() throws CatalogException {
+		if (client == null) {
+			client = getMetastoreClient(hiveConf);
+			LOG.info("Connected to Hive metastore");
+		}
+	}
+
+	@Override
+	public void close() throws CatalogException {
+		if (client != null) {
+			client.close();
+			client = null;
+			LOG.info("Close connection to Hive metastore");
+		}
+	}
+
+	// ------ databases ------
+
+	@Override
+	public String getCurrentDatabase() throws CatalogException {
+		return currentDatabase;
+	}
+
+	@Override
+	public void setCurrentDatabase(String databaseName) throws DatabaseNotExistException, CatalogException {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(databaseName));
+
+		if (!databaseExists(databaseName)) {
+			throw new DatabaseNotExistException(catalogName, databaseName);
+		}
+
+		currentDatabase = databaseName;
+	}
+
+	@Override
+	public List<String> listDatabases() throws CatalogException {
+		try {
+			return client.getAllDatabases();
+		} catch (TException e) {
+			throw new CatalogException(
+				String.format("Failed to list all databases in %s", catalogName), e);
+		}
+	}
+
+	@Override
+	public CatalogDatabase getDatabase(String databaseName) throws DatabaseNotExistException, CatalogException {
+		Database hiveDb;
+
+		try {
+			hiveDb = client.getDatabase(databaseName);
+		} catch (NoSuchObjectException e) {
+			throw new DatabaseNotExistException(catalogName, databaseName);
+		} catch (TException e) {
+			throw new CatalogException(
+				String.format("Failed to get database %s from %s", databaseName, catalogName), e);
+		}
+
+		return new GenericCatalogDatabase(hiveDb.getParameters(), hiveDb.getDescription());
+	}
+
+	@Override
+	public boolean databaseExists(String databaseName) throws CatalogException {
+		try {
+			return client.getDatabase(databaseName) != null;
```

Comment:
what is different return null and throw `NoSuchObjectException`  of `getDatabase`?
Use try `getDatabase` catch `DatabaseNotExistException` directly?

Reply:
`client.getDatabase(databaseName) != null;` is used only for the purpose to return a boolean, which will be `true` if the`x  != null` gets to run, for this method. `client.getDatabase()` will throw `NoSuchObjectException` if the database doesn't exist, and we catch that to return a `false`.